src/IO/ReadBufferFromRemoteFS.h

```diff
@@ -0,0 +1,46 @@
+#pragma once
+
+#include <Disks/IDiskRemote.h>
+#include <IO/ReadBufferFromFile.h>
+
+
+namespace DB
+{
+
+class ReadBufferFromRemoteFS : public ReadBufferFromFileBase
+{
+friend class ThreadPoolRemoteFSReader;
+
+public:
+    explicit ReadBufferFromRemoteFS(const RemoteMetadata & metadata_);
+
+    off_t seek(off_t offset, int whence) override;
+
+    off_t getPosition() override { return absolute_position - available(); }
+
+    String getFileName() const override { return metadata.metadata_file_path; }
+
+    void reset();
+
+protected:
+    size_t readInto(char * data, size_t size, size_t offset);
+
+    virtual SeekableReadBufferPtr createReadBuffer(const String & path) const = 0;
```

Comment:
What does it mean? Why buffer need to create another buffer?

Reply:
ReadBufferFromRemoteFS has children (s3, hdfs, web) which implement this virtual method and return a reader. We use this reader to read from s3, hdfs, web respectively.